### PR TITLE
feat(bigquery): add support for optimized query path

### PR DESCRIPTION
**Pull request by shollyman**
For users who invoke Read() directly on a Query object,
this change evaluates the query config and runs the query
using an optimized path when possible.  In cases where it's
not possible to run using the optimized path, we fall back
to the existing mechanism which uses jobs.insert and polling.

This PR builds on the RowIterator changes in #2484 which
also allow us to construct an iterator with a cache of the first
page of results, which further reduces latency by avoiding an
extra network round trip.

Generally, this improves small query result scenarios, but as resultsize grows the performance difference converges.  If you measure time until Read() returns this can appear to be a less performant option, as some of the cost of assembling results is done before the call returns (first page caching).  We also find that tail latency variance can still be quite high.

For internal users (this doc is inaccessible to general public), specific measurements here: https://docs.google.com/spreadsheets/d/1zyaGVdcj-nDj10QAXinoddgarNkwH6345cb7F8zyJIk/edit?usp=sharing

### CODE REVIEW

File: bigquery/query.go [hunk 1]
```diff
@@ -324,11 +326,97 @@ func (q *Query) newJob() (*bq.Job, error) {
...
+		Query:              q.QueryConfig.Q,
+		Location:           q.Location,
+		UseLegacySql:       &pfalse,
+		MaximumBytesBilled: q.QueryConfig.MaxBytesBilled,
+		RequestId:          uid.NewSpace("request", nil).New(),
+		Labels:             q.Labels,
+	}
+	if q.QueryConfig.DisableQueryCache {
+		pfalse := false
+		qRequest.UseQueryCache = &pfalse
```
Comment (hongalex) [inline bigquery/query.go:405]: Is there a reason why you need a new false pointer here as opposed to reusing pfalse from L394?

File: bigquery/bigquery.go [hunk 2]
```diff
@@ -109,6 +109,28 @@ func (c *Client) insertJob(ctx context.Context, job *bq.Job, media io.Reader) (*
...
+	setClientHeader(call.Header())
+
+	var res *bq.QueryResponse
+	var err error
+	invoke := func() error {
+		res, err = call.Do()
+		return err
+	}
+
+	// we expect to control request ID, so we can always runWithRetry
```
Comment (hongalex) [inline bigquery/bigquery.go:126]: nit: should be a sentence.

File: bigquery/query.go (hunk 1 above)
Comment (shollyman) [reply bigquery/query.go:404]: Hah, good spot!  Corrected.

File: bigquery/bigquery.go (hunk 2 above)
Comment (shollyman) [reply bigquery/bigquery.go]: Done.

**Pull request merged**